File: externals/libbot-drc/bot2-lcmgl/java/src/bot_lcmgl/LCMGL.java
package bot_lcmgl;

import java.io.*;

import lcm.lcm.LCM;

public class LCMGL 
{
    ByteArrayOutputStream _bouts = new ByteArrayOutputStream();
    DataOutputStream _outs = new DataOutputStream(_bouts);
    String _channel = "LCMGL";
    LCM _lcm = null;
    String _name = "unnamed";
    int _scene = 0;
    int _sequence = 0;

    final static int LCMGL_BEGIN         = 4;
    final static int LCMGL_END           = 5;
    final static int LCMGL_VERTEX3F      = 6;
    final static int LCMGL_VERTEX3D      = 7;
    final static int LCMGL_COLOR3F       = 8;
    final static int LCMGL_COLOR4F       = 9;
    final static int LCMGL_POINTSIZE     = 10;
    final static int LCMGL_ENABLE        = 11;
    final static int LCMGL_DISABLE       = 12;
    final static int LCMGL_BOX           = 13;
    final static int LCMGL_CIRCLE        = 14;
    final static int LCMGL_LINE_WIDTH    = 15;
    final static int LCMGL_NOP           = 16;
    final static int LCMGL_VERTEX2D      = 17;
    final static int LCMGL_VERTEX2F      = 18;
    final static int LCMGL_TEXT          = 19;
    final static int LCMGL_DISK          = 20;
    final static int LCMGL_TRANSLATED    = 21;
    final static int LCMGL_ROTATED       = 22;
    final static int LCMGL_LOAD_IDENTITY = 23;
    final static int LCMGL_PUSH_MATRIX   = 24;
    final static int LCMGL_POP_MATRIX    = 25;
    final static int LCMGL_RECT          = 26;
    final static int LCMGL_TEXT_LONG     = 27;
    final static int LCMGL_NORMAL3F      = 28;
    final static int LCMGL_SCALEF        = 29;
    final static int LCMGL_MULT_MATRIXF  = 30;
    final static int LCMGL_MULT_MATRIXD  = 31;
    final static int LCMGL_MATERIALF     = 32;

    public LCMGL(LCM lcm, String name)
    {
        _lcm = lcm;
        _name = name;
    }

    private synchronized void add0(int cmd)
    {
        try { _outs.writeByte(cmd); } catch(IOException xcp) {}
    }

    private synchronized void add1i(int cmd, int arg)
    {
        try { 
            _outs.writeByte(cmd); 
            _outs.writeInt(arg);
        } catch(IOException xcp) {}
    }

    private synchronized void add1f(int cmd, float arg)
    {
        try { 
            _outs.writeByte(cmd); 
            _outs.writeFloat(arg);
        } catch(IOException xcp) {}
    }

    private synchronized void add2f(int cmd, float x, float y)
    {
        try { 
            _outs.writeByte(cmd); 
            _outs.writeFloat(x);
            _outs.writeFloat(y);
        } catch(IOException xcp) {}
    }

    private synchronized void add2d(int cmd, double x, double y)
    {
        try { 
            _outs.writeByte(cmd); 
            _outs.writeDouble(x);
            _outs.writeDouble(y);
        } catch(IOException xcp) {}
    }

    private synchronized void add3f(int cmd, float x, float y, float z)
    {
        try { 
            _outs.writeByte(cmd); 
            _outs.writeFloat(x);
            _outs.writeFloat(y);
            _outs.writeFloat(z);
        } catch(IOException xcp) {}
    }

    private synchronized void add3d(int cmd, double x, double y, double z)
    {
        try { 
            _outs.writeByte(cmd); 
            _outs.writeDouble(x);
            _outs.writeDouble(y);
            _outs.writeDouble(z);
        } catch(IOException xcp) {}
    }

    private synchronized void add4f(int cmd, float x, float y, float z, float w)
    {
        try { 
            _outs.writeByte(cmd); 
            _outs.writeFloat(x);
            _outs.writeFloat(y);
            _outs.writeFloat(z);
            _outs.writeFloat(w);
        } catch(IOException xcp) {}
    }

    private synchronized void add4d(int cmd, double x, double y, double z, double w)
    {
        try { 
            _outs.writeByte(cmd); 
            _outs.writeDouble(x);
            _outs.writeDouble(y);
            _outs.writeDouble(z);
            _outs.writeDouble(w);
        } catch(IOException xcp) {}
    }

    public synchronized void switchBuffers()
    {
        byte[] b = _bouts.toByteArray();
        bot_lcmgl.data_t msg = new bot_lcmgl.data_t();
        msg.name = _name;
        msg.scene = _scene;
        msg.sequence = _sequence;
        msg.data = b;
        msg.datalen = b.length;

        _lcm.publish(_channel, msg);

        _scene++;
        _sequence++;
        _bouts.reset();
    }

    public synchronized void glBegin(int mode) { add1i(LCMGL_BEGIN, mode); }
    public synchronized void glEnd() { add0(LCMGL_END); }

    public synchronized void glEnable(int cap) { add1i(LCMGL_ENABLE, cap); }
    public synchronized void glDisable(int cap) { add1i(LCMGL_DISABLE, cap); }

    public synchronized void glPushMatrix() { add0(LCMGL_PUSH_MATRIX); }
    public synchronized void glPopMatrix() { add0(LCMGL_POP_MATRIX); }
    public synchronized void glLoadIdentity() { add0(LCMGL_LOAD_IDENTITY); }
    public synchronized void glMultMatrixf(float f[]) {
        try {
            _outs.writeByte(LCMGL_MULT_MATRIXF);
            for(int i=0; i<16; i++)
                _outs.writeFloat(f[i]);
        } catch(IOException xcp) {}
    }
    public synchronized void glMultMatrixd(double f[]) {
        try {
            _outs.writeByte(LCMGL_MULT_MATRIXF);
            for(int i=0; i<16; i++)
                _outs.writeDouble(f[i]);
        } catch(IOException xcp) {}
    }

    public synchronized void glPointSize(float size) { add1f(LCMGL_POINTSIZE, size); }
    public synchronized void glLineWidth(float size) { add1f(LCMGL_LINE_WIDTH, size); }

    public synchronized void glColor3f(float r, float g, float b) { add3f(LCMGL_COLOR3F, r, g, b); }
    public synchronized void glColor4f(float r, float g, float b, float a) { add4f(LCMGL_COLOR4F, r, g, b, a); }

    public synchronized void glScalef(float x, float y, float z) { add3f(LCMGL_SCALEF, x, y, z); }
    public synchronized void glTranslated(double x, double y, double z) { add3d(LCMGL_TRANSLATED, x, y, z); }
    public synchronized void glRotated(double theta, double x, double y, double z) { add4d(LCMGL_ROTATED, theta, x, y, z); }

    public synchronized void glNormal3f(float x, float y, float z) { add3f(LCMGL_NORMAL3F, x, y, z); }

    public synchronized void glVertex2d(double x, double y) { add2d(LCMGL_VERTEX2D, x, y); }
    public synchronized void glVertex2f(float x, float y) { add2f(LCMGL_VERTEX2F, x, y); }

    public synchronized void glVertex3d(double x, double y, double z) { add3d(LCMGL_VERTEX3D, x, y, z); }
    public synchronized void glVertex3f(float x, float y, float z) { add3f(LCMGL_VERTEX3F, x, y, z); }

    public synchronized void circle(double x, double y, double z, double r) {
        try {
            _outs.writeByte(LCMGL_CIRCLE); 
            _outs.writeDouble(x);
            _outs.writeDouble(y);
            _outs.writeDouble(z);
            _outs.writeFloat((float)r);
        } catch(IOException xcp) {}
    }

    public synchronized void disk(double x, double y, double z, double r_in, double r_out) {
        try {
            _outs.writeByte(LCMGL_DISK); 
            _outs.writeDouble(x);
            _outs.writeDouble(y);
            _outs.writeDouble(z);
            _outs.writeFloat((float)r_in);
            _outs.writeFloat((float)r_out);
        } catch(IOException xcp) {}
    }
    
    public static final int TEXT_DROP_SHADOW = 1;
    public static final int TEXT_JUSTIFY_LEFT = 2;
    public static final int TEXT_JUSTIFY_RIGHT = 4;
    public static final int TEXT_JUSTIFY_CENTER = 8;
    public static final int TEXT_ANCHOR_LEFT = 16;
    public static final int TEXT_ANCHOR_RIGHT = 32;
    public static final int TEXT_ANCHOR_TOP = 64;
    public static final int TEXT_ANCHOR_BOTTOM = 128;
    public static final int TEXT_ANCHOR_HCENTER = 256;
    public static final int TEXT_ANCHOR_VCENTER = 512;
    public static final int TEXT_NORMALIZED_SCREEN_COORDINATES = 1024;
    public static final int TEXT_MONOSPACED = 2048;    
    
    public synchronized void text(double x, double y, double z, String text) {
        this.text(x, y, z, text, 0, 
                TEXT_DROP_SHADOW | 
                TEXT_JUSTIFY_CENTER |
                TEXT_ANCHOR_HCENTER |
                TEXT_ANCHOR_VCENTER);
    }
    
    public synchronized void text(double x, double y, double z, String text,
            int font, int flags) {
        try {
            _outs.writeByte(LCMGL_TEXT_LONG);
            _outs.writeInt(font);
            _outs.writeInt(flags);
            _outs.writeDouble(x);
            _outs.writeDouble(y);
            _outs.writeDouble(z);
            _outs.writeInt(text.length());
            _outs.writeBytes(text);
        } catch(IOException xcp) {}
    }

    public static void main(String args[])
    {
        LCM lcm = LCM.getSingleton();
        LCMGL lcmgl = new LCMGL(lcm, "Java test");

        // constants taken from javax.media.opengl.GL
        int GL_LINE_LOOP = 2; 
        int GL_LINES = 1;

        double theta = 0;
        double x = 0;
        double y = 0;
        double z = 0;
        double t = 0;
        while(true) {
            lcmgl.glPushMatrix();
            lcmgl.glRotated(theta, 0, 0, 1);
            lcmgl.glTranslated(x, y, z);
            for(int scale=1; scale<=2; scale++) {
                lcmgl.glPushMatrix();
                lcmgl.glScalef(scale, scale, scale);

                lcmgl.glColor3f(scale * 0.5f, scale * 0.5f, scale * 0.5f);

                lcmgl.glBegin(GL_LINE_LOOP);
                lcmgl.glVertex3f(-1, -1, -1);
                lcmgl.glVertex3f(-1,  1, -1);
                lcmgl.glVertex3f( 1,  1, -1);
                lcmgl.glVertex3f( 1, -1, -1);
                lcmgl.glEnd();

                lcmgl.glBegin(GL_LINE_LOOP);
                lcmgl.glVertex3f(-1, -1,  1);
                lcmgl.glVertex3f(-1,  1,  1);
                lcmgl.glVertex3f( 1,  1,  1);
                lcmgl.glVertex3f( 1, -1,  1);
                lcmgl.glEnd();

                lcmgl.glBegin(GL_LINES);
                lcmgl.glVertex3f(-1, -1, -1);
                lcmgl.glVertex3f(-1, -1,  1);
                lcmgl.glVertex3f(-1,  1, -1);
                lcmgl.glVertex3f(-1,  1,  1);
                lcmgl.glVertex3f( 1,  1, -1);
                lcmgl.glVertex3f( 1,  1,  1);
                lcmgl.glVertex3f( 1, -1, -1);
                lcmgl.glVertex3f( 1, -1,  1);
                lcmgl.glEnd();

                lcmgl.glPopMatrix();
            }
            lcmgl.glPopMatrix();
            lcmgl.switchBuffers();

            try { Thread.sleep(30); } catch (InterruptedException xcp) { }
            t += 0.1;
            x = Math.cos(t);
            y = Math.sin(t);
            z = 0;
            theta = 0.1 * t * 180 / Math.PI;
        }
    }

}

File: externals/libbot-drc/bot2-lcmgl/python/src/bot_lcmgl/lcmgl.py
import cStringIO as StringIO
import bot_lcmgl.data_t as data_t
import struct

LCMGL_GL_BEGIN         = 4
LCMGL_GL_END           = 5
LCMGL_GL_VERTEX3F      = 6
LCMGL_GL_VERTEX3D      = 7
LCMGL_GL_COLOR3F       = 8
LCMGL_GL_COLOR4F       = 9
LCMGL_GL_POINTSIZE     = 10
LCMGL_GL_ENABLE        = 11
LCMGL_GL_DISABLE       = 12
LCMGL_BOX              = 13  # nyi
LCMGL_CIRCLE           = 14
LCMGL_GL_LINE_WIDTH    = 15
LCMGL_NOP              = 16
LCMGL_GL_VERTEX2D      = 17
LCMGL_GL_VERTEX2F      = 18
LCMGL_TEXT             = 19
LCMGL_DISK             = 20
LCMGL_GL_TRANSLATED    = 21
LCMGL_GL_ROTATED       = 22
LCMGL_GL_LOAD_IDENTITY = 23
LCMGL_GL_PUSH_MATRIX   = 24
LCMGL_GL_POP_MATRIX    = 25
LCMGL_RECT             = 26  # nyi
LCMGL_TEXT_LONG        = 27
LCMGL_GL_NORMAL3F      = 28
LCMGL_GL_SCALEF        = 29
LCMGL_GL_MULT_MATRIXF  = 30  # nyi
LCMGL_GL_MULT_MATRIXD  = 31  # nyi
LCMGL_GL_MATERIALF     = 32
LCMGL_GL_PUSH_ATTRIB   = 33  # nyi
LCMGL_GL_POP_ATTRIB    = 34  # nyi
LCMGL_GL_DEPTH_FUNC    = 35  # nyi
LCMGL_TEXTURE2D        = 36
LCMGL_TEXTURE_DRAW_QUAD= 37
LCMGL_SPHERE           = 38
LCMGL_CYLINDER         = 39

# text flags
LCMGL_TEXT_DROP_SHADOW                   = 1
LCMGL_TEXT_JUSTIFY_LEFT                  = 2
LCMGL_TEXT_JUSTIFY_RIGHT                 = 4
LCMGL_TEXT_JUSTIFY_CENTER                = 8
LCMGL_TEXT_ANCHOR_LEFT                   = 16
LCMGL_TEXT_ANCHOR_RIGHT                  = 32
LCMGL_TEXT_ANCHOR_TOP                    = 64
LCMGL_TEXT_ANCHOR_BOTTOM                 = 128
LCMGL_TEXT_ANCHOR_HCENTER                = 256
LCMGL_TEXT_ANCHOR_VCENTER                = 512
LCMGL_TEXT_NORMALIZED_SCREEN_COORDINATES = 1024
LCMGL_TEXT_MONOSPACED                    = 2048

# texture constants
LCMGL_LUMINANCE = 0
LCMGL_RGB = 1
LCMGL_RGBA = 2

LCMGL_COMPRESS_NONE = 0

# redefine OpenGL constants
GL_POINTS         = 0x0000
GL_LINES          = 0x0001
GL_LINE_LOOP      = 0x0002
GL_LINE_STRIP     = 0x0003
GL_TRIANGLES      = 0x0004
GL_TRIANGLE_STRIP = 0x0005
GL_TRIANGLE_FAN   = 0x0006
GL_QUADS          = 0x0007
GL_QUAD_STRIP     = 0x0008
GL_POLYGON        = 0x0009

GL_NEVER             = 0x0200
GL_LESS              = 0x0201
GL_EQUAL             = 0x0202
GL_LEQUAL            = 0x0203
GL_GREATER           = 0x0204
GL_NOTEQUAL          = 0x0205
GL_GEQUAL            = 0x0206
GL_ALWAYS            = 0x0207
GL_DEPTH_TEST        = 0x0B71
GL_DEPTH_BITS        = 0x0D56
GL_DEPTH_CLEAR_VALUE = 0x0B73
GL_DEPTH_FUNC        = 0x0B74
GL_DEPTH_RANGE       = 0x0B70
GL_DEPTH_WRITEMASK   = 0x0B72
GL_DEPTH_COMPONENT   = 0x1902

GL_LIGHTING = 0xb50

GL_FRONT = 0x0404
GL_BACK = 0x0405

GL_AMBIENT = 0x1200
GL_DIFFUSE = 0x1201
GL_SPECULAR = 0x1202
GL_SHININESS = 0x1601
GL_EMISSION = 0x1600

for n in range(1, 9):
    args = ", ".join(["a%d" % i for i in range(n)])
    exec """
def _lcmgl_make_encode_%d(cmd, fmt):
    st = struct.Struct(">B%%s" %% fmt)
    def encode(self, %s):
        self.data.write(st.pack(cmd, %s))
        self.datalen += st.size
    return encode
""" % (n, args, args)

def _lcmgl_make_encode_0(cmd):
    data = struct.pack("B", cmd)
    def encode(self):
        self.data.write(data)
        self.datalen += 1
    return encode
    
class lcmgl:
    def __init__(self, name, lcm):
        self.lcm = lcm
        self.data = StringIO.StringIO()
        self.datalen = 0
        self.scene = 1
        self.name = name
        self.ntextures = 0

    def switch_buffer(self):
        d = self.data.getvalue()

        msg = data_t()
        msg.name = self.name
        msg.scene = self.scene
        msg.sequence = 0
        msg.datalen = len(d)
        msg.data = d
        self.lcm.publish("LCMGL", msg.encode())

        self.ntextures = 0
        self.data = StringIO.StringIO()
        self.scene += 1

    glBegin        = _lcmgl_make_encode_1(LCMGL_GL_BEGIN, "I")
    glEnd          = _lcmgl_make_encode_0(LCMGL_GL_END)

    glVertex2f     = _lcmgl_make_encode_2(LCMGL_GL_VERTEX2F, "ff")
    glVertex2d     = _lcmgl_make_encode_2(LCMGL_GL_VERTEX2D, "dd")
    glVertex3f     = _lcmgl_make_encode_3(LCMGL_GL_VERTEX3F, "fff")
    glVertex3d     = _lcmgl_make_encode_3(LCMGL_GL_VERTEX3D, "ddd")

    glColor3f      = _lcmgl_make_encode_3(LCMGL_GL_COLOR3F, "fff")
    glColor4f      = _lcmgl_make_encode_4(LCMGL_GL_COLOR4F, "ffff")

    glPointSize    = _lcmgl_make_encode_1(LCMGL_GL_POINTSIZE, "f")
    glLineWidth    = _lcmgl_make_encode_1(LCMGL_GL_LINE_WIDTH, "f")

    glEnable       = _lcmgl_make_encode_1(LCMGL_GL_ENABLE, "I")
    glDisable      = _lcmgl_make_encode_1(LCMGL_GL_DISABLE, "I")

    glTranslated   = _lcmgl_make_encode_3(LCMGL_GL_TRANSLATED, "ddd")
    glRotated      = _lcmgl_make_encode_4(LCMGL_GL_ROTATED, "dddd")
    glScalef       = _lcmgl_make_encode_3(LCMGL_GL_SCALEF, "fff")

    glLoadIdentity = _lcmgl_make_encode_0(LCMGL_GL_LOAD_IDENTITY)
    glPushMatrix   = _lcmgl_make_encode_0(LCMGL_GL_PUSH_MATRIX)
    glPopMatrix    = _lcmgl_make_encode_0(LCMGL_GL_POP_MATRIX)

    glNormal3f     = _lcmgl_make_encode_3(LCMGL_GL_NORMAL3F, "fff")
    glMaterialf    = _lcmgl_make_encode_6(LCMGL_GL_MATERIALF, "IIffff")

    # circle(x, y, z, radius)
    circle         = _lcmgl_make_encode_4(LCMGL_CIRCLE, "dddf")
    # disk(x, y, z, r_in, r_out)
    disk           = _lcmgl_make_encode_5(LCMGL_DISK, "dddff")
    # sphere(x, y, z, radius, slices, stacks)
    sphere         = _lcmgl_make_encode_6(LCMGL_SPHERE, "ddddII")
    # cylinder(x, y, z, r_base, r_top, height, slices, stacks)
    cylinder       = _lcmgl_make_encode_8(LCMGL_CYLINDER, "ddddddII")

    def text(self, x, y, z, text, flags = 0):
        font = 0
        self.data.write(struct.pack(">BIIdddI", LCMGL_TEXT_LONG, font, flags, x, y, z, len(text)))
        self.data.write(text)
        self.datalen += 37 + len(text)

    def texture2d(self, data, width, height, format, compression):
        self.ntextures += 1
        tex_id = self.ntextures

        if format not in [ LCMGL_LUMINANCE, LCMGL_RGB, LCMGL_RGBA ]:
            raise ValueError("Invalid format")
#        bytes_per_pixel = 1
#        if format == LCMGL_LUMINANCE:
#            bytes_per_pixel = 1
#        elif format == LCMGL_RGB:
#            bytes_per_pixel = 3
#        elif format == LCMGL_RGBA:
#            bytes_per_pixel = 4
#        else:
#        bytes_per_row = width * bytes_per_pixel
#        datalen = bytes_per_row * height

        if compression == LCMGL_COMPRESS_NONE:
            data_tosend = data
        else:
            raise ValueError("Invalid compression value")

        self.data.write(struct.pack(">BIIIIII", LCMGL_TEXTURE2D, tex_id,
            width, height, format, compression, len(data_tosend)))
        self.data.write(data_tosend)

        return tex_id

    def textureDrawQuad(self, tex_id, 
            top_left_xyz, bot_left_xyz,
            bot_right_xyz, top_right_xyz):

        if tex_id > self.ntextures or tex_id <= 0:
            raise ValueError("Invalid texture ID")
        self.data.write(struct.pack(">BIdddddddddddd", LCMGL_TEXTURE_DRAW_QUAD,
            tex_id,
            top_left_xyz[0], top_left_xyz[1], top_left_xyz[2], 
            bot_left_xyz[0], bot_left_xyz[1], bot_left_xyz[2], 
            bot_right_xyz[0], bot_right_xyz[1], bot_right_xyz[2], 
            top_right_xyz[0], top_right_xyz[1], top_right_xyz[2]))


if __name__ == "__main__":
    import array
    import math
    import lcm
    lcm = lcm.LCM()
    a = array.array('B')
    width = 100
    height = 100
    g = lcmgl('lcmgl_texture', lcm)
    for y in range(height):
        for x in range(width):
            v = math.sin(x / 5.) + math.cos(y / 5.)
            a.append(int(v * 50 + 127))
    img_data = a.tostring()
    print len(img_data)
    tex_id = g.texture2d(img_data, width, height, LCMGL_LUMINANCE, LCMGL_COMPRESS_NONE)
    g.glColor3f(0, 0, 1)
    g.textureDrawQuad(tex_id, 
            (-10, 10, 0),
            (-10, -10, 0),
            (10, -10, 0),
            (10, 10, 0))
    g.switch_buffer()

    import time
    g = lcmgl('lcmgl_test', lcm)
    for i in range(90):
        g.glLineWidth(1 + i / 25.0)
        g.glColor3f(1, 1, 0)

        px, py = 0, 0
        nx, ny = 0, 0

        g.glBegin(GL_LINES)
        for j in range(i):
            nx = 0.1 * j * math.cos(j / 5.)
            ny = 0.1 * j * math.sin(j / 5.)
            g.glVertex3d(px, py, 0)
            g.glVertex3d(nx, ny, 0)
            px, py = nx, ny
        g.glEnd()

        g.glEnable(GL_LIGHTING)
        g.glMaterialf(GL_FRONT, GL_AMBIENT, 0, 0.7, 0, 1);
        g.glMaterialf(GL_FRONT, GL_DIFFUSE, 0, 0.7, 0, 1);

        g.sphere(nx, ny, 0, 1, 10, 10)

        g.switch_buffer()
        time.sleep(0.05)
